File: app/src/main/java/pl/srw/billcalculator/wrapper/HistoryRepo.java
package pl.srw.billcalculator.wrapper;

import org.greenrobot.greendao.query.LazyList;

import javax.inject.Inject;
import javax.inject.Singleton;

import pl.srw.billcalculator.db.History;
import pl.srw.billcalculator.persistence.Database;
import pl.srw.billcalculator.persistence.type.BillType;

@Singleton
public class HistoryRepo {

    @Inject
    HistoryRepo() {
    }

    public void deleteBillWithPrices(BillType billType, Long id, Long pricesId) {
        Database.deleteBillWithPrices(billType, id, pricesId);
    }

    public void undoDelete() {
        Database.undelete();
    }

    public LazyList<History> getAll() {
        return Database.getHistory(); // TODO: Rewrite to Rx
    }
}

File: app/src/main/java/pl/srw/billcalculator/wrapper/HistoryRepo.kt
package pl.srw.billcalculator.wrapper

import org.greenrobot.greendao.query.LazyList
import pl.srw.billcalculator.db.Bill
import pl.srw.billcalculator.db.History
import pl.srw.billcalculator.db.Prices
import pl.srw.billcalculator.persistence.Database
import pl.srw.billcalculator.persistence.type.BillType
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HistoryRepo @Inject internal constructor() {

    private var deleted: Collection<Pair<Bill, Prices>> = emptyList()

    // TODO: Rewrite to Rx
    fun getAll(): LazyList<History> = Database.getHistory()

    fun deleteBillWithPrices(bill: Bill) {
        Database.deleteBillWithPrices(BillType.valueOf(bill), bill.id, bill.pricesId)
    }

    fun cacheBillsForUndoDelete(bills: Collection<Bill>) {
        val prices = bills.map { loadPricesForBill(it)}
        deleted = bills.zip(prices)
    }

    fun cacheBillForUndoDelete(bill: Bill) {
        deleted = listOf(Pair(bill, loadPricesForBill(bill)))
    }

    fun undoDelete() {
        for ((bill, prices) in deleted) {
            Database.insert(bill, prices)
        }
        deleted = emptyList()
    }

    fun isUndoDeletePossible()= !deleted.isEmpty()

    private fun loadPricesForBill(bill: Bill): Prices {
        val billType = BillType.valueOf(bill)
        val loadPrices = Database.loadPrices(billType, bill.pricesId)
        return checkNotNull(loadPrices, {"Prices not found for id ${bill.pricesId}"})
    }
}
